src/app/cli/src/commands/system_diagnose_command.rs

```diff
@@ -0,0 +1,197 @@
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fs::File;
+use std::io::{Stdout, Write};
+use std::path::Path;
+use std::sync::Arc;
+
+use console::style;
+use container_runtime::ContainerRuntime;
+use futures::TryStreamExt;
+use kamu::domain::{DatasetRepository, VerificationOptions, VerificationService};
+
+use super::{CLIError, Command};
+
+pub const DUMMY_IMAGE: &str = "docker.io/busybox:latest";
+
+pub struct SystemDiagnoseCommand {
+    dataset_repo: Arc<dyn DatasetRepository>,
+    verification_svc: Arc<dyn VerificationService>,
+}
+
+impl SystemDiagnoseCommand {
+    pub fn new(
+        dataset_repo: Arc<dyn DatasetRepository>,
+        verification_svc: Arc<dyn VerificationService>,
+    ) -> Self {
+        Self {
+            dataset_repo,
+            verification_svc,
+        }
+    }
+}
+
+#[async_trait::async_trait(?Send)]
+impl Command for SystemDiagnoseCommand {
+    async fn run(&mut self) -> Result<(), CLIError> {
+        SystemDiagnose::check(self.dataset_repo.clone(), self.verification_svc.clone()).await?;
+        Ok(())
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct SystemDiagnose {}
+
+impl SystemDiagnose {
+    pub async fn check(
+        dataset_repo: Arc<dyn DatasetRepository>,
+        verification_svc: Arc<dyn VerificationService>,
+    ) -> Result<(), CLIError> {
+        let diagnose_check = RunCheck::new();
+        diagnose_check.run(dataset_repo, verification_svc).await
+    }
+}
+
+pub struct RunCheck {
+    container_runtime: ContainerRuntime,
+}
+
+impl RunCheck {
+    fn new() -> Self {
+        Self {
+            container_runtime: ContainerRuntime::default(),
+        }
+    }
+
+    async fn run_container_check(&self, f: &mut Stdout) -> Result<(), CLIError> {
+        write!(f, "container installed... ")?;
+
+        let is_container_installed = match self
+            .container_runtime
+            .custom_cmd("--version".to_string())
+            .output()
+            .await
+        {
+            Ok(_) => style("ok").green(),
+            Err(_) => style("failed").red(),
+        };
+
+        write!(f, "{is_container_installed}")?;
+        Ok(())
+    }
+
+    async fn run_container_pull_check(&self, f: &mut Stdout) -> Result<(), CLIError> {
+        write!(f, "\ncontainer can pull images... ")?;
+
+        let is_container_image_pulled =
+            match self.container_runtime.pull_image(DUMMY_IMAGE, None).await {
+                Ok(_) => style("ok").green(),
+                Err(_) => style("failed").red(),
+            };
+
+        write!(f, "{is_container_image_pulled}")?;
+        Ok(())
+    }
+
+    async fn run_container_rootless_check(&self, f: &mut Stdout) -> Result<(), CLIError> {
+        write!(f, "\ncontainer rootless check... ")?;
```

Comment:
Will previous `pull images` test even pass without a rootless property properly configured? If no, then move it up.

Reply:
Changed a bit logic, to check is it possible to run a container. About rootles this chack will fail even on the first step